[media] Create AudioDiscardDurationTracker

osagie98 opened the pull request:
Creates a class AudioDiscardDurationTracker to store the discard durations of audio buffers sent to pass-through audio renderers. The stored durations can then be used during media time calculation. This is implemented for the Android pass-through audio renderer.

b/313719729

osagie98 committed:
[media] Create AudioDiscardDurationTracker

Creates a class AudioDiscardDurationTracker to store the discard
durations of audio buffers sent to pass-through audio renderers. The
stored durations can then be used during media time calculation. This
is implemented for the Android pass-through audio renderer.

b/313719729

jasonzhangxx commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,76 @@
+  const AudioSampleInfo audio_sample_info = input_buffer.audio_sample_info();
+  discard_durations_by_timestamp_[input_buffer.timestamp()] =
+      audio_sample_info.discarded_duration_from_front +
+      audio_sample_info.discarded_duration_from_back;
```
I think we can pause the media time when it reaches input buffer's timestamp (if `discarded_duration_from_front` is not zero) until the media time exceeds input's timestamp + `discarded_duration_from_front`. And for `discarded_duration_from_back`, we can move or combine it to `discarded_duration_from_front` of next input buffer. What do you think?

osagie98 replied:
I like the approach with `discarded_duration_from_front`. With `discarded_duration_from_back`, ideally I'd like to pause media time from <input timestamp - `discarded_duration_from_back`> to <input timestamp + buffer duration>, but I need to know the length of the buffer. We don't know the length of the buffer until we read the E/AC3 sync frame, however I know it's likely 32 ms (1536 samples @ 48 khz). I'm working to come up with a different solution.

osagie98 replied:
I ended up hardcoding 32 ms and adding a `TODO` to address later.

osagie98 committed:
Only modify media time once frames are played

osagie98 commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,76 @@
+void AudioDiscardDurationTracker::CacheMultipleDiscardDurations(
+    const InputBuffers& input_buffers) {
+  for (const auto& input_buffer : input_buffers) {
+    CacheDiscardDuration(*input_buffer.get());
```
Done

osagie98 committed:
Add AudioDiscardDurationTracker tests

This commit also adds a passthrough only option to
GetSupportedAudioTestFiles(), as well as refines the
AudioDiscardDurationTracker implementation.

osagie98 committed:
Filter tests from unsupported platforms

osagie98 committed:
Allow uninstantiated parameterized test

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h:
```diff
@@ -0,0 +1,103 @@
+  // the length of the buffer, a single AudioDiscardInfo is created to discard
+  // the entire length of the buffer.
+  void CacheDiscardDuration(const scoped_refptr<InputBuffer>& input_buffer,
+                            SbTime buffer_length);
```
buffer_length => buffer_duration

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,101 @@
+    // As a lot of time may have passed since the last call to
+    // AdjustTimeForTotalDiscardDuration(), remove all AudioDiscardInfos that
+    // have already been passed by the |timestamp|.
+    while (discard_infos_.size() > 0 &&
+           timestamp >= discard_infos_.front().discard_end_timestamp) {
+      total_discard_duration_ += discard_infos_.front().discard_duration;
+      discard_infos_.pop();
+    }
```
After the loop, `timestamp` may fall into one of the discarded range, and should be processed as what's done between line 77 to line 82.  Considering run this loop first, and then the check for timestamp falls in a discarded range.

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,101 @@
+using shared::starboard::media::AudioSampleInfo;
+
+void AudioDiscardDurationTracker::CacheDiscardDuration(
+    const scoped_refptr<InputBuffer>& input_buffer,
```
Just to double check, if there is an audio input buffer with absolute time range 24.95 seconds to 25.05 seconds.  Now we are seeking to 25.00 seconds, which may form a partial audio input buffer with 50 ms discarded from the front.  In such case, is the timestamp of the input buffer passed in already adjusted to 25.00 seconds, or still 24.95 seconds?

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h:
```diff
@@ -0,0 +1,103 @@
+    SbTime discard_start_timestamp;
+    SbTime discard_end_timestamp;
+  };
+
```
Consider adding ThreadChecker here (a lesson learnt from b/289281412).

osagie98 replied:
I added a mutex since `AdjustTimeForTotalDiscardDuration()` is called from `GetCurrentMediaTime()`, so it's expected to be called on a different thread than the other methods.

osagie98 committed:
Add mutex to AudioDiscardDurationTracker

xiaomings commented on starboard/android/shared/audio_renderer_passthrough.cc:
```diff
@@ -138,6 +138,11 @@ void AudioRendererPassthrough::WriteSamples(const InputBuffers& input_buffers) {
       std::bind(&AudioRendererPassthrough::OnDecoderConsumed, this));
+  // AC-3/EAC-3 sync frames with 1536 samples at 48khz have a length of 32 ms.
+  // TODO: Determine this value at runtime.
+  const SbTime kBufferLength = 32 * kSbTimeMillisecond;
```
nit: kBufferDuration

Also consider moving the call to CacheMultipleDiscardDurations() before the call to decoder_->Decode() to avoid potential race conditions.

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h:
```diff
@@ -0,0 +1,106 @@
+
+ private:
+  struct AudioDiscardInfo {
+    SbTime discard_duration;
```
if `end_timestamp = start_timestamp + duration` is always true, we don't have to store all of them.  Consider storing the start and end timestamps.

xiaomings commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,106 @@
+  SbTime discard_duration_from_back =
+      input_buffer->audio_sample_info().discarded_duration_from_back;
+
+  ScopedLock lock(mutex_);
```
This function should be private, and the lock should be held inside CacheMultipleDiscardDurations(), so we can ensure that the effect of calling CacheMultipleDiscardDurations() is consistent to the observers.

osagie98 committed:
Make CacheDiscardDuration Private

osagie98 committed:
Check for mutex acquisition

jasonzhangxx commented on starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc:
```diff
@@ -0,0 +1,107 @@
+      << ". AudioDiscardDurationTracker::AdjustTimeForTotalDiscardDuration() "
+         "requires monotonically increasing timestamps.";
+  last_received_timestamp_ = timestamp;
+
```
`discard_start_timestamp` is based on adjusted timestamp, but the input `timestamp` is not adjusted. It should be adjusted before comparing to the `discard_start_timestamp`. And the return value is `discard_start_timestamp - total_discard_duration_`, that adjusts the timestamp on an adjusted timestamp.

For example, if there're two inputs and both have 20ms audio and discarded duration from front of 5ms, the total played time would be 40ms(wall time), and `discard_infos_` would be {0ms, 5ms}, {15ms, 5ms}. During 0-15ms (wall time), the returned value works as expected. During 15-20ms(wall time), the returned value would be 10ms, which is not right. It's expected to freeze the timestamp at 20-25ms(wall time).

osagie98 committed:
Merge branch 'main' into address-passthrough-partial-audio

osagie98 committed:
Remove use of SbTime

osagie98 commented on starboard/android/shared/audio_renderer_passthrough.cc:
```diff
@@ -135,6 +135,12 @@ void AudioRendererPassthrough::WriteSamples(const InputBuffers& input_buffers) {
+  // AC-3/EAC-3 sync frames with 1536 samples at 48khz have a duration of 32 ms.
+  // TODO: Determine this value at runtime.
+  const SbTime kBufferDuration = 32 * kSbTimeMillisecond;
+  discard_duration_tracker_.CacheMultipleDiscardDurations(input_buffers,
```
Or rather, just the discard durations would need to be stored.

osagie98 committed:
Fix double time adjustment issue
